contracts/Strategy.sol

```diff
@@ -53,80 +143,224 @@ contract Strategy is BaseStrategy {
             uint256 _debtPayment
         )
     {
-        // TODO: Do stuff here to free up any returns back into `want`
-        // NOTE: Return `_profit` which is value generated by all positions, priced in `want`
-        // NOTE: Should try to free up at least `_debtOutstanding` of underlying position
+        (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+        if (block.timestamp > lastAccrued) {
+            kashiPair.accrue();
+        }
+
+        uint256 assets = estimatedTotalAssets();
+        uint256 wantBal = want.balanceOf(address(this));
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
+
+        if (assets >= debt) {
+            _debtPayment = _debtOutstanding;
+            _profit = assets.sub(debt);
+
+            uint256 amountToFree = _profit.add(_debtPayment);
+
+            if (amountToFree > 0 && wantBal < amountToFree) {
+                liquidatePosition(amountToFree);
+
+                uint256 newLoose = want.balanceOf(address(this));
+
+                //if we dont have enough money adjust _debtOutstanding and only change profit if needed
+                if (newLoose < amountToFree) {
+                    if (_profit > newLoose) {
+                        _profit = newLoose;
+                        _debtPayment = 0;
+                    } else {
+                        _debtPayment = Math.min(
+                            newLoose.sub(_profit),
+                            _debtPayment
+                        );
+                    }
+                }
+            }
+        } else {
+            //serious loss should never happen but if it does lets record it accurately
+            _loss = debt.sub(assets);
+        }
     }
 
     function adjustPosition(uint256 _debtOutstanding) internal override {
-        // TODO: Do something to invest excess `want` tokens (from the Vault) into your positions
-        // NOTE: Try to adjust positions so that `_debtOutstanding` can be freed up on *next* harvest (not immediately)
+        if (emergencyExit) {
+            return;
+        }
+
+        uint256 wantBalance = want.balanceOf(address(this));
+
+        uint256 shares = 0;
+
+        if (wantBalance > 0) {
```

Comment:
it might be worth it to add a DUST_THRESHOLD to avoid depositing dust each time, making the harvest / tend more expensive for low amounts.